### ISSUE TITLE
opencv-3.1.0/modules/core/src/opengl.cpp:1644:9: error: use of undeclared identifier 'clGetGLContextInfoKHR_fn'

### ISSUE THREAD

**Issue by JohnTian** [Itseez/opencv#6052]
```cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D PYTHON2_PACKAGES_PATH=~/.virtualenvs/cv/lib/python2.7/site-packages -D PYTHON2_LIBRARY=/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/bin -D PYTHON2_INCLUDE_DIR=/usr/local/Frameworks/Python.framework/Headers -D INSTALL_C_EXAMPLES=OFF -D INSTALL_PYTHON_EXAMPLES=ON -D BUILD_EXAMPLES=ON -D OPENCV_EXTRA_MODULES_PATH=~/bin/opencv/opencv_contrib-3.1.0/modules -D WITH_TBB=ON -D WITH_QT=ON -D WITH_OPENGL=ON -D WITH_V4L=ON ..```

I have an issue like:
```
/opencv-3.1.0/modules/core/src/opengl.cpp:1644:9: error: use of undeclared identifier 'clGetGLContextInfoKHR_fn'; did you mean
      'clGetContextInfo'?
        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
        ^~~~~~~~~~~~~~~~~~~~~~~~
        clGetContextInfo
/System/Library/Frameworks/OpenCL.framework/Headers/cl.h:633:1: note: 'clGetContextInfo' declared here
clGetContextInfo(cl_context         /* context */,
^
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1644:33: error: expected ';' after expression
        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
                                ^
                                ;
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1644:34: error: use of undeclared identifier 'clGetGLContextInfoKHR'
        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
                                 ^
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1644:59: error: use of undeclared identifier 'clGetGLContextInfoKHR_fn'; did you mean
      'clGetContextInfo'?
        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
                                                          ^~~~~~~~~~~~~~~~~~~~~~~~
                                                          clGetContextInfo
/System/Library/Frameworks/OpenCL.framework/Headers/cl.h:633:1: note: 'clGetContextInfo' declared here
clGetContextInfo(cl_context         /* context */,
^
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1646:14: error: use of undeclared identifier 'clGetGLContextInfoKHR'; did you mean 'clGetContextInfo'?
        if (!clGetGLContextInfoKHR)
             ^~~~~~~~~~~~~~~~~~~~~
             clGetContextInfo
/System/Library/Frameworks/OpenCL.framework/Headers/cl.h:633:1: note: 'clGetContextInfo' declared here
clGetContextInfo(cl_context         /* context */,
^
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1646:14: error: address of function 'clGetContextInfo' will always evaluate to 'true'
      [-Werror,-Wpointer-bool-conversion]
        if (!clGetGLContextInfoKHR)
            ~^~~~~~~~~~~~~~~~~~~~~
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1646:14: note: prefix with the address-of operator to silence this warning
        if (!clGetGLContextInfoKHR)
             ^
             &
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1669:52: error: use of undeclared identifier 'CL_CURRENT_DEVICE_FOR_GL_CONTEXT_KHR'
        status = clGetGLContextInfoKHR(properties, CL_CURRENT_DEVICE_FOR_GL_CONTEXT_KHR, sizeof(cl_device_id), (void*)&device, NULL);
                                                   ^
/Users/tianzhaixing/bin/opencv/opencv-3.1.0/modules/core/src/opengl.cpp:1644:9: warning: expression result unused [-Wunused-value]
        clGetGLContextInfoKHR_fn clGetGLContextInfoKHR = (clGetGLContextInfoKHR_fn)
        ^~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 7 errors generated.
make[2]: *** [modules/core/CMakeFiles/opencv_core.dir/src/opengl.cpp.o] Error 1
```

**Comment by JohnTian** [comment]
Just now, I remove **-D WITH_OPENGL=ON**. It works ! **But not the final solution.**
Following is how i do it.
1. **mbp:build tianzhaixing$ cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D PYTHON2_PACKAGES_PATH=~/.virtualenvs/cv/lib/python2.7/site-packages -D PYTHON2_LIBRARY=/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/bin -D PYTHON2_INCLUDE_DIR=/usr/local/Frameworks/Python.framework/Headers -D INSTALL_C_EXAMPLES=OFF -D INSTALL_PYTHON_EXAMPLES=ON -D BUILD_EXAMPLES=ON -D OPENCV_EXTRA_MODULES_PATH=~/bin/opencv/opencv_contrib-3.1.0/modules -D WITH_TBB=ON -D WITH_QT=ON  -D WITH_V4L=ON ..**  
2. **mbp:build tianzhaixing$ make -j8**  
3. **mbp:build tianzhaixing$ sudo make install**

**Comment by alalek** [comment]
Please add output of this command (from build directory):
```
grep OPENCL cvconfig.h
```

**Comment by JohnTian** [comment]
@alalek 
```
mbp:build tianzhaixing$ grep OPENCL cvconfig.h
#define HAVE_OPENCL
#define HAVE_OPENCL_STATIC
/* #undef HAVE_OPENCL_SVM */
```

**Comment by alalek** [comment]
`HAVE_OPENCL_STATIC` means that used system OpenCL headers (via `#include <OpenCL/cl_gl.h>`).
```
grep clGetGLContextInfoKHR_fn /System/Library/Frameworks/OpenCL.framework/Headers/cl_gl.h
```
Is output empty on your machine?

**Comment by jmbapps** [comment]
The framework is already built for iOS, download, add to project.  I have a vid HowTo for iOS.  For OSX you are out of luck, I havent been able to get it to work yet.  Bool conflicts with Apple macro.

**Comment by ahundt** [comment]
This bug seems to have been reproduced by a number of people on OS X:
https://github.com/Homebrew/homebrew-science/issues/3147

**Comment by JohnTian** [comment]
@alalek Sorry for something else. Yes, I got empty after run 
```
grep clGetGLContextInfoKHR_fn /System/Library/Frameworks/OpenCL.framework/Headers/cl_gl.h
```

**Comment by stromnov** [comment]
OS X support has been stalled at OpenCL 1.2 since the release of OS X Mavericks. But clGetGLContextInfoKHR_fn requires OpenCL 2.0+.

**Comment by alalek** [comment]
Not exactly.
It is available in Khronos public headers since [OpenCL 1.0](https://www.khronos.org/registry/cl/api/1.0/cl_gl.h), but it is optional.

**Comment by stromnov** [comment]
@alalek yes, you right, sorry for misinformation. Anyway, OpenCL from OS X don't support that extension.

**Comment by ahundt** [comment]
Why is this closed now? Seems a bot closed it but I'm still fairly certain a workaround is needed

**Comment by mshabunin** [comment]
@ahundt , it has been closed because of merging #6060. Does it fix the issue?

**Comment by ahundt** [comment]
@mshabunin missed that, thanks! I went back to the latest 2.x OpenCV so can't test trivially.
